core/capabilities/ccip/ccipsolana/executecodec.go

```diff
@@ -0,0 +1,259 @@
+	if len(report.ChainReports) == 0 || len(report.ChainReports) > 1 {
+		return nil, fmt.Errorf("unexpected chain report length: %d", len(report.ChainReports))
+	}
+
+	var message ccip_router.Any2SolanaRampMessage
+	chainReport := report.ChainReports[0]
+	if len(chainReport.Messages) > 0 {
+		// currently only allow committing one message at a time
+		msg := chainReport.Messages[0]
+		tokenAmounts := make([]ccip_router.Any2SolanaTokenTransfer, 0, len(msg.TokenAmounts))
+		for _, tokenAmount := range msg.TokenAmounts {
+			if tokenAmount.Amount.IsEmpty() {
+				return nil, fmt.Errorf("empty amount for token: %s", tokenAmount.DestTokenAddress)
+			}
+
+			if len(tokenAmount.DestTokenAddress) != solana.PublicKeyLength {
+				return nil, fmt.Errorf("invalid destTokenAddress address: %v", tokenAmount.DestTokenAddress)
+			}
+
+			tokenAmounts = append(tokenAmounts, ccip_router.Any2SolanaTokenTransfer{
+				SourcePoolAddress: tokenAmount.SourcePoolAddress,
+				DestTokenAddress:  solana.PublicKeyFromBytes(tokenAmount.DestTokenAddress),
+				ExtraData:         tokenAmount.ExtraData,
+				Amount:            bigIntToBytes32(tokenAmount.Amount),
+				DestGasAmount:     bytesToUint32(tokenAmount.DestExecData),
+			})
+		}
+
+		family, err := chainsel.GetSelectorFamily(uint64(chainReport.SourceChainSelector))
+		if err != nil {
+			return nil, fmt.Errorf("invalid source chain selector: %w", err)
+		}
+
+		// if source chain is Solana the Borsh decoder will be used, for EVM we will construct the extra args from
+		// the chain agnostic extra args map
+		var extraArgs ccip_router.SolanaExtraArgs
+		switch family {
+		case chainsel.FamilySolana:
+			decoder := agbinary.NewBorshDecoder(msg.ExtraArgs)
+			err = extraArgs.UnmarshalWithDecoder(decoder)
+			if err != nil {
+				return nil, fmt.Errorf("invalid extra arguments: %w", err)
+			}
+		case chainsel.FamilyEVM:
+			extraArgs, err = parseExtraArgsMap(msg.ExtraArgsDecoded)
+			if err != nil {
+				return nil, fmt.Errorf("invalid extra args map: %w", err)
+			}
+		default:
+			return nil, fmt.Errorf("unsupported source chain family: %s", family)
+		}
+
+		if len(msg.Receiver) != solana.PublicKeyLength {
+			return nil, fmt.Errorf("invalid receiver address: %v", msg.Receiver)
+		}
+
+		message = ccip_router.Any2SolanaRampMessage{
+			Header: ccip_router.RampMessageHeader{
+				MessageId:           msg.Header.MessageID,
+				SourceChainSelector: uint64(msg.Header.SourceChainSelector),
+				DestChainSelector:   uint64(msg.Header.DestChainSelector),
+				SequenceNumber:      uint64(msg.Header.SequenceNumber),
+				Nonce:               msg.Header.Nonce,
+			},
+			Sender:       msg.Sender,
+			Data:         msg.Data,
+			Receiver:     solana.PublicKeyFromBytes(msg.Receiver),
+			TokenAmounts: tokenAmounts,
+			ExtraArgs:    extraArgs,
+		}
+	}
+
+	var offchainTokenData [][]byte
+	if len(chainReport.OffchainTokenData) > 0 {
+		offchainTokenData = chainReport.OffchainTokenData[0]
+	}
```

Comment:
We should only have an offchain token data if there are tokens as part of the message, perhaps that can be validated